MaterialLib/MPL/mpMedium.cpp

```diff
@@ -0,0 +1,143 @@
+/**
+ * \author Norbert Grunwald
+ * \date   07.09.2017
+ * \brief
+ *
+ * \copyright
+ * Copyright (c) 2012-2017, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include "mpMedium.h"
+#include <boost/variant.hpp>
+#include <iostream>
+#include <string>
+#include "Properties/pAverageVolumeFraction.h"
+#include "Properties/pConstant.h"
+#include "mpComponent.h"
+#include "mpPhase.h"
+
+namespace MaterialPropertyLib
+{
+/**
+ * This constructor parses the "phases" and "properties" subtrees
+ * of the config tree and calls create methods for the phase vector
+ * and the properties array. Medium properties are optional. If not
+ * defined, default properties are assigned.
+ */
+Medium::Medium(BaseLib::ConfigTree const& config)
+{
+    // Default properties are initialized in the first place, such
+    // that user-defined properties overwrite those defaults.
+    createDefaultProperties();
+
+    // A name of a medium is entirely optional and has no effects
+    // other than a small gain of clarity in case several media
+    // should be defined.
+    auto const medium_name =
+        config.getConfigParameter<std::string>("name", "no_name");
+
+    _properties[name] = std::make_unique<Constant>(medium_name);
+
+    // Parsing the phases
+    auto const phases_config = config.getConfigSubtree("phases");
+    createPhases(phases_config);
+    // Parsing medium properties, overwriting the defaults.
+    auto const properties_config =
+        config.getConfigSubtreeOptional("properties");
+    if (properties_config)
+        createProperties(properties_config.get());
```

Comment:
style: boost::optional has an `operator*()`, so you could also write `createProperties(*properties_config)`.